v3/src/liftoff/core.ts

```diff
@@ -0,0 +1,101 @@
+import { trace, Bond } from './pattern'
+import { Ref, isRef, DEFAULT_BOND, getLabel } from './ref'
+import { throws } from './errors'
+import { getLocation } from './loc'
+
+export type Plan = () => void
+
+export class Core {
+  private defs: Map<Ref<any>, Set<Bond>> = new Map
+
+  constructor(public readonly plan: Plan) {
+    this.link(plan)
+  }
+
+  link(plan: Plan) {
+    const {defs} = this
+    for (const delta of trace(plan)) {
+      const {mut, bond} = delta
+      if (mut === 'add') {
+        if (bond.type === 'def') {
+          const {state: {ref}} = bond
+          if (!defs.has(ref))
+            defs.set(ref, new Set([bond]))
+          else
+            defs.get(ref)!.add(bond)
+        }
+      }
+      // TODO (queerviolet): Handle change and delete deltas
+    }
+  }
+
+  states: Map<Ref<any>, any> = new Map
+
+  /**
+   * Resolve all bonds that define this ref.
+   *
+   * @param ref {Ref<T>} the ref
+   * @returns Set<Bond>
+   */
+  private resolveDefs(ref: Ref<any>): Set<Bond> {
+    return this.defs.get(ref) || (
+      typeof ref[DEFAULT_BOND] !== 'undefined' ? new Set([ref[DEFAULT_BOND]]) : NO_BONDS
+    )
+  }
+
+  /**
+   * Given a ref, return an array of all defined values for that ref.
+   *
+   * @param ref {Ref<T>} the ref to read
+   * @returns Promise<T[]>
+   */
+  async once<T>(ref: Ref<T>): Promise<T[]> {
+    const {states} = this
```

Comment:
Mmm, I tend to do this whenever I'm going to use `this.something` multiple times in a method. I like the rhythm it gives the code (all those `this.`s can get tiresome), and I have some suspicion it might be verrrry slightly faster, by virtue of avoiding the lookup into this? But that motivation is quite tiny.